File: tools/sweep_config_test.py
#!/usr/bin/env python3
"""Tests for sweep_config — the YAML/JSON sweep-profile loader/saver.

These exercise the PURE (de)serialization surface (load_profile / save_profile /
get_profile_path / list_profiles) against a tmp dir, so the round-trip invariant
— save_profile then load_profile recovers every field, including nested models
and price hints — is proven without touching a real sweep. The defaults a
half-specified YAML falls back to are pinned too, since the loader's `.get(...,
default)` calls are the contract a profile author relies on."""
import json
import tempfile
import unittest
from pathlib import Path

import sweep_config as sc


def sample_profile():
    return sc.SweepProfile(
        name="nightly",
        description="the nightly sweep",
        models=[
            sc.ModelConfig(
                name="zai/glm-4.7-flash",
                provider="zai",
                base_url="https://api.example/v1",
                api_key_env="ZAI_KEY",
                price_hint=sc.PriceHint(input=0.5, output=1.5, source="docs"),
                enabled=True,
            ),
            sc.ModelConfig(name="local/qwen", provider="local",
                           local_shim="shims/qwen.sh", enabled=False),
        ],
        workload=sc.WorkloadConfig(max_turns=20, trials=3, timeout_s=900,
                                   transcript_path="t/x.jsonl"),
        output_dir="out/here",
        skip_api=True,
        tags=["a", "b"],
        public=False,
    )


class RoundTrip(unittest.TestCase):
    def test_save_then_load_recovers_every_field(self):
        p = sample_profile()
        with tempfile.TemporaryDirectory() as d:
            path = Path(d) / "nightly.yaml"
            sc.save_profile(p, path)
            got = sc.load_profile(path)

        self.assertEqual(got.name, p.name)
        self.assertEqual(got.description, p.description)
        self.assertEqual(got.output_dir, p.output_dir)
        self.assertEqual(got.skip_api, p.skip_api)
        self.assertEqual(got.tags, p.tags)
        self.assertEqual(got.public, p.public)
        # workload
        self.assertEqual(got.workload.max_turns, 20)
        self.assertEqual(got.workload.trials, 3)
        self.assertEqual(got.workload.timeout_s, 900)
        self.assertEqual(got.workload.transcript_path, "t/x.jsonl")
        # models, in order
        self.assertEqual([m.name for m in got.models],
                         ["zai/glm-4.7-flash", "local/qwen"])
        m0, m1 = got.models
        self.assertEqual(m0.provider, "zai")
        self.assertEqual(m0.base_url, "https://api.example/v1")
        self.assertEqual(m0.api_key_env, "ZAI_KEY")
        self.assertTrue(m0.enabled)
        self.assertIsNotNone(m0.price_hint)
        self.assertEqual(m0.price_hint.input, 0.5)
        self.assertEqual(m0.price_hint.output, 1.5)
        self.assertEqual(m0.price_hint.source, "docs")
        self.assertFalse(m1.enabled)
        self.assertEqual(m1.local_shim, "shims/qwen.sh")
        # the second model carried no price hint, so it must come back as None
        self.assertIsNone(m1.price_hint)


class Defaults(unittest.TestCase):
    def test_minimal_json_profile_uses_documented_defaults(self):
        # A profile with only a name must load with every documented default.
        with tempfile.TemporaryDirectory() as d:
            path = Path(d) / "bare.json"
            path.write_text(json.dumps({"name": "bare"}), encoding="utf-8")
            got = sc.load_profile(path)
        self.assertEqual(got.name, "bare")
        self.assertEqual(got.description, "")
        self.assertEqual(got.models, [])
        self.assertEqual(got.output_dir, "fak/experiments/agent-live/sweep")
        self.assertFalse(got.skip_api)
        self.assertTrue(got.public)
        # workload defaults
        self.assertEqual(got.workload.max_turns, 12)
        self.assertEqual(got.workload.trials, 1)
        self.assertEqual(got.workload.timeout_s, 600)
        self.assertIsNone(got.workload.transcript_path)

    def test_model_provider_defaults_to_unknown(self):
        with tempfile.TemporaryDirectory() as d:
            path = Path(d) / "m.json"
            path.write_text(json.dumps({"name": "p", "models": [{"name": "x"}]}),
                            encoding="utf-8")
            got = sc.load_profile(path)
        self.assertEqual(len(got.models), 1)
        self.assertEqual(got.models[0].provider, "unknown")
        self.assertTrue(got.models[0].enabled)
        self.assertIsNone(got.models[0].price_hint)


class ProfilePaths(unittest.TestCase):
    def test_get_profile_path_prefers_existing_yaml(self):
        with tempfile.TemporaryDirectory() as d:
            dd = Path(d)
            (dd / "p.yml").write_text("name: p\n", encoding="utf-8")
            # only .yml exists -> returns it
            self.assertEqual(sc.get_profile_path("p", dd), dd / "p.yml")
            # now a .yaml exists too -> .yaml wins (checked first)
            (dd / "p.yaml").write_text("name: p\n", encoding="utf-8")
            self.assertEqual(sc.get_profile_path("p", dd), dd / "p.yaml")

    def test_get_profile_path_defaults_to_yaml_when_absent(self):
        with tempfile.TemporaryDirectory() as d:
            dd = Path(d)
            self.assertEqual(sc.get_profile_path("nope", dd), dd / "nope.yaml")

    def test_list_profiles_loads_all_in_dir(self):
        with tempfile.TemporaryDirectory() as d:
            dd = Path(d)
            sc.save_profile(sc.SweepProfile(name="one"), dd / "one.yaml")
            sc.save_profile(sc.SweepProfile(name="two"), dd / "two.yaml")
            names = sorted(p.name for p in sc.list_profiles(dd))
            self.assertEqual(names, ["one", "two"])


if __name__ == "__main__":
    unittest.main()

File: internal/sweepconfig/sweepconfig_test.go
package sweepconfig

import (
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"sort"
	"strconv"
	"strings"
	"testing"
)

func sampleProfile() SweepProfile {
	return SweepProfile{
		Name:        "nightly",
		Description: "the nightly sweep",
		Models: []ModelConfig{
			{
				Name:      "zai/glm-4.7-flash",
				Provider:  "zai",
				BaseURL:   "https://api.example/v1",
				APIKeyEnv: "ZAI_KEY",
				PriceHint: &PriceHint{Input: 0.5, Output: 1.5, Source: "docs"},
				Enabled:   true,
			},
			{Name: "local/qwen", Provider: "local", LocalShim: "shims/qwen.sh", Enabled: false},
		},
		Workload:  WorkloadConfig{MaxTurns: 20, Trials: 3, TimeoutS: 900, TranscriptPath: "t/x.jsonl"},
		OutputDir: "out/here",
		SkipAPI:   true,
		Tags:      []string{"a", "b"},
		Public:    false,
	}
}

func TestSaveThenLoadRecoversEveryField(t *testing.T) {
	p := sampleProfile()
	path := filepath.Join(t.TempDir(), "nightly.yaml")
	if err := SaveProfile(p, path); err != nil {
		t.Fatal(err)
	}
	got, err := LoadProfile(path)
	if err != nil {
		t.Fatal(err)
	}
	if !reflect.DeepEqual(got, p) {
		t.Fatalf("profile mismatch\ngot:  %+v\nwant: %+v", got, p)
	}
}

func TestMinimalJSONProfileUsesDocumentedDefaults(t *testing.T) {
	path := filepath.Join(t.TempDir(), "bare.json")
	b, _ := json.Marshal(map[string]any{"name": "bare"})
	if err := osWrite(path, b); err != nil {
		t.Fatal(err)
	}
	got, err := LoadProfile(path)
	if err != nil {
		t.Fatal(err)
	}
	if got.Name != "bare" || got.Description != "" || len(got.Models) != 0 || got.OutputDir != "fak/experiments/agent-live/sweep" || got.SkipAPI || !got.Public {
		t.Fatalf("defaults wrong: %+v", got)
	}
	if got.Workload.MaxTurns != 12 || got.Workload.Trials != 1 || got.Workload.TimeoutS != 600 || got.Workload.TranscriptPath != "" {
		t.Fatalf("workload defaults wrong: %+v", got.Workload)
	}
}

func TestModelProviderDefaultsToUnknown(t *testing.T) {
	path := filepath.Join(t.TempDir(), "m.json")
	b, _ := json.Marshal(map[string]any{"name": "p", "models": []any{map[string]any{"name": "x"}}})
	if err := osWrite(path, b); err != nil {
		t.Fatal(err)
	}
	got, err := LoadProfile(path)
	if err != nil {
		t.Fatal(err)
	}
	if len(got.Models) != 1 || got.Models[0].Provider != "unknown" || !got.Models[0].Enabled || got.Models[0].PriceHint != nil {
		t.Fatalf("model defaults wrong: %+v", got.Models)
	}
}

func TestProfilePaths(t *testing.T) {
	dir := t.TempDir()
	if err := osWrite(filepath.Join(dir, "p.yml"), []byte("name: p\n")); err != nil {
		t.Fatal(err)
	}
	if got := GetProfilePath("p", dir); got != filepath.Join(dir, "p.yml") {
		t.Fatalf("path = %q", got)
	}
	if err := osWrite(filepath.Join(dir, "p.yaml"), []byte("name: p\n")); err != nil {
		t.Fatal(err)
	}
	if got := GetProfilePath("p", dir); got != filepath.Join(dir, "p.yaml") {
		t.Fatalf("path = %q", got)
	}
	if got := GetProfilePath("nope", dir); got != filepath.Join(dir, "nope.yaml") {
		t.Fatalf("absent path = %q", got)
	}
}

func TestListProfilesLoadsAllInDir(t *testing.T) {
	dir := t.TempDir()
	if err := SaveProfile(DefaultProfile("one"), filepath.Join(dir, "one.yaml")); err != nil {
		t.Fatal(err)
	}
	if err := SaveProfile(DefaultProfile("two"), filepath.Join(dir, "two.yaml")); err != nil {
		t.Fatal(err)
	}
	var names []string
	for _, p := range ListProfiles(dir) {
		names = append(names, p.Name)
	}
	sort.Strings(names)
	if !reflect.DeepEqual(names, []string{"one", "two"}) {
		t.Fatalf("names = %+v", names)
	}
}

func TestExistingSimpleYAMLShapeLoads(t *testing.T) {
	path := filepath.Join(t.TempDir(), "quick.yaml")
	body := []byte("name: quick\nworkload:\n  max_turns: 5\nmodels:\n  - name: m\n    provider: p\n    price_hint:\n      input: 0.07\n      output: 0.40\n      source: docs\ntags:\n  - smoke\npublic: true\n")
	if err := osWrite(path, body); err != nil {
		t.Fatal(err)
	}
	got, err := LoadProfile(path)
	if err != nil {
		t.Fatal(err)
	}
	if got.Name != "quick" || got.Workload.MaxTurns != 5 || len(got.Models) != 1 || got.Models[0].PriceHint == nil || got.Tags[0] != "smoke" {
		t.Fatalf("parsed yaml wrong: %+v", got)
	}
}

func TestMalformedOrUnsupportedYAMLReturnsLineAwareError(t *testing.T) {
	tests := []struct {
		name string
		body string
		line int
	}{
		{name: "missing colon", body: "name: valid\nthis is not yaml\n", line: 2},
		{name: "block scalar", body: "name: valid\ndescription: |\n  unsupported\n", line: 2},
		{name: "flow sequence", body: "name: valid\ntags: [one, two]\n", line: 2},
		{name: "wrong indentation", body: "name: valid\nworkload:\n   max_turns: 5\n", line: 3},
		{name: "unknown key", body: "name: valid\nfuture_option: value\n", line: 2},
		{name: "unterminated quote", body: "name: valid\ndescription: \"unfinished\n", line: 2},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			path := filepath.Join(t.TempDir(), "bad.yaml")
			if err := osWrite(path, []byte(tt.body)); err != nil {
				t.Fatal(err)
			}
			_, err := LoadProfile(path)
			if err == nil {
				t.Fatal("LoadProfile succeeded for malformed or unsupported YAML")
			}
			want := "line " + strconv.Itoa(tt.line)
			if !strings.Contains(err.Error(), want) {
				t.Fatalf("error %q does not contain %q", err, want)
			}
		})
	}
}

func TestQuotedScalarsAndInlineCommentsLoad(t *testing.T) {
	path := filepath.Join(t.TempDir(), "quoted.yaml")
	body := []byte(`name: "quick: smoke #1" # profile comment
description: 'It''s # literal: yes' # trailing comment
models:
  - name: "model:#1" # trailing comment
    provider: 'provider #1'
    base_url: "https://example.test/v1?q=a#frag" # trailing comment
    api_key_env: "KEY:VALUE"
    local_shim: 'C:\tools\shim #1'
    price_hint:
      input: 0.5 # trailing comment
      output: 1.25 # trailing comment
      source: 'docs #1'
    enabled: true # trailing comment
workload:
  max_turns: 5 # trailing comment
  trials: 2
  timeout_s: 30
  transcript_path: " transcripts/run #1.jsonl " # spaces are data
output_dir: ' out: #1 '
skip_api: false # trailing comment
skip_offline: true
skip_local_shim: false
fail_fast: true
tags:
  - "smoke #1" # trailing comment
  - 'it''s: quoted'
  - owner's-tag
public: false # trailing comment
`)
	if err := osWrite(path, body); err != nil {
		t.Fatal(err)
	}
	got, err := LoadProfile(path)
	if err != nil {
		t.Fatal(err)
	}
	want := SweepProfile{
		Name:        "quick: smoke #1",
		Description: "It's # literal: yes",
		Models: []ModelConfig{{
			Name:      "model:#1",
			Provider:  "provider #1",
			BaseURL:   "https://example.test/v1?q=a#frag",
			APIKeyEnv: "KEY:VALUE",
			LocalShim: `C:\tools\shim #1`,
			PriceHint: &PriceHint{Input: 0.5, Output: 1.25, Source: "docs #1"},
			Enabled:   true,
		}},
		Workload:    WorkloadConfig{MaxTurns: 5, Trials: 2, TimeoutS: 30, TranscriptPath: " transcripts/run #1.jsonl "},
		OutputDir:   " out: #1 ",
		SkipOffline: true,
		FailFast:    true,
		Tags:        []string{"smoke #1", "it's: quoted", "owner's-tag"},
		Public:      false,
	}
	if !reflect.DeepEqual(got, want) {
		t.Fatalf("profile mismatch\ngot:  %+v\nwant: %+v", got, want)
	}
}

func TestSaveThenLoadPreservesYAMLIndicatorStrings(t *testing.T) {
	special := ` leading: # "quote" \ trailing `
	p := SweepProfile{
		Name:        special,
		Description: special,
		Models: []ModelConfig{{
			Name:      special,
			Provider:  special,
			BaseURL:   special,
			APIKeyEnv: special,
			LocalShim: special,
			PriceHint: &PriceHint{Input: 0.5, Output: 1.5, Source: special},
			Enabled:   true,
		}},
		Workload:      WorkloadConfig{MaxTurns: 7, Trials: 2, TimeoutS: 45, TranscriptPath: special},
		OutputDir:     special,
		SkipAPI:       true,
		SkipOffline:   true,
		SkipLocalShim: true,
		FailFast:      true,
		Tags:          []string{special},
		Public:        true,
	}
	path := filepath.Join(t.TempDir(), "special.yaml")
	if err := SaveProfile(p, path); err != nil {
		t.Fatal(err)
	}
	got, err := LoadProfile(path)
	if err != nil {
		t.Fatal(err)
	}
	if !reflect.DeepEqual(got, p) {
		t.Fatalf("profile mismatch\ngot:  %+v\nwant: %+v", got, p)
	}
}

func osWrite(path string, b []byte) error {
	return os.WriteFile(path, b, 0o644)
}
